docs/protocol.md

```diff
@@ -0,0 +1,333 @@
+
+Each of these 3 networks has their own genesis account, block and amount. Blocks
+intended for one network will not be accepted on another network, as they will
+fail to validate. XXX is this true?
+
+The VersionMax, VersionUsing and VersionMin values are related to the version
+of the network protocol being used. The VersionUsing value indicates the protocol
+version a packet is adhering to.
+
+The message Type can have one of the following values (see `message_type` in `rai/node/common.hpp`):
+
+  | Type       | Value |
+  | ---------- | ----- |
+  | Keepalive  | 2     |
+  | Publish    | 3     |
+  | ConfirmReq | 4     |
+  | ConfirmAck | 5     |
+
+The Extensions field currently only contains the 8-bit block type, in the highest
+8 bits of the field. But this is only the case if the message contains a Block (i.e. Publish, ConfirmReq and ConfirmAck messages).
+See `rai::message::block_type_set()` in `rai/node/common.cpp`.
+
+## Keepalive message
+
+Reference: class `rai::keepalive` in `rai/node/common.cpp`
+
+A Keepalive message contains a list of IP addresses of peers, in the form of (IPv6 address, port) pairs.
+Each IP address is stored as 16 bytes, each port as 2 bytes:
+
+  | Field        | Size (bytes) | Description     |
+  | ------------ | ------------ | --------------- |
+  | IPv6 address | 16           | Big-endian?     |
+  | Port         | 2            | Little-endian?! |
+
+The number of pairs in the message can vary, from 1 up to 8 peers.
+
+## Publish message
+
+Reference: class `rai::publish` in `rai/node/common.cpp`
+
+A Publish message simply contains a published Block, which was either created or
+forwarded by another node.
+
+See below for the way a block in encoded.
+
+## ConfirmReq message
+
+Reference: class `rai::confirm_req` in `rai/node/common.cpp`
+
+XXX A ConfirmReq message contains a Block for which a vote has been requested?
+
+  | Field | Size (bytes) | Description        |
+  | ----- | ------------ | ------------------ |
+  | Block | Varies       | A serialized Block |
+
+## ConfirmAck message
+
+Reference: class `rai::confirm_ack` in `rai/node/common.cpp`
+
+A ConfirmAck message contains a Vote by a representative on a Block.
+
+  | Field | Size (bytes) | Description       |
+  | ----- | ------------ | ----------------- |    
+  | Vote  | (varies)     | A serialized Vote |
+
+
+## Vote
+
+Reference: class `rai::vote` in `rai/common.cpp`
+
+  | Field     | Size (bytes) | Description                          |
+  | --------- | ------------ | ------------------------------------ |
+  | Account   | 32           | Public key of the voting Account     |
+  | Signature | 64           | Signature                            |
+  | Sequence  | 8            | 64-bit sequence number (big-endian?) |
+  | Block     | Varies       | The Block being voted on             |
+
+The Signature of a Vote signs the following list of hashables: `hash(Block)`, `Sequence`
+
+XXX what is the exact function of the sequence number?
```

Comment:
A representative can change its vote for a block in a fork (i.e. if the other block won). This is to ensure that the new vote overrides the old vote.

Reply:
Was explaining this yesterday to someone; an example:

Say for example, my node originally votes for block X; I send sequence number of 0 and my vote signature for block X. Now, when another node is processing my votes, it sees my vote for X and adds my weight to X in that election, and that my vote had sequence number 0. Later, my node observes > 50% of voting power is now voting for the opposite block and so now I want to change my vote. Well now I vote for block Y and attach a sequence number of 1. Other nodes then see my vote for Y, compare the new sequence number to the sequence number it's already stored for my rep and sees that the new sequence is greater (1 > 0), so it now replaces my vote and adds my stake to block Y. The reason this is required is because without the sequence, it's possible that my vote for block Y could reach this node first and my vote for block X could have reached it second. In that case, without the sequence number, it would then think my latest vote was for block X, which is of course the opposite of what it should be.